### ISSUE TITLE
Chrome browser hang on large collection

### ISSUE THREAD

**Issue by waz1500** [gotson/komga#86]
### Komga environment
- **Linux/Docker**: 
- **Xeon CPU E3-1230V3 with 32GB RAM**: 
- [x] **I am running Komga with Docker**
  - Docker image tag [v0.13.1]: 
- [x] **I have a problem in the web interface**
  - Chrome Version 79.0.3945.130 (Official Build) (64-bit): 

### Describe the bug
Using Komga on my main collection of 7449 books is painfully slow. Browser becomes unresponsive. Thumbnail book cover take minutes to load and only on a few books. Very slow response from browser. Unable to click on anything.   

### Steps to reproduce
1. Add library to komga. 
2. Let it analyze books. My collection is large so left it running for a few days. 
3. Log into Komga and see that it has detected all books. 
4. Thumbnail cover for all books seem to be generating live when using the browser.  

### Expected behavior
Able to use Komga with Chrome and not have it hang.   

### Additional context
I know you mentioned somewhere that you also have a large-ish collection. I'm not too sure why it hangs for me but you dont have issues with your own collection. My collection is all in zip files and on the HDD. I'm not using any cloud service etc. 

It seems like Komga only generates thumbnails when the user is browsing the Komga interface. Is it not possible for Komga to make thumbnails for content somewhere and have it load up when browsing.   

### Log file
```
2020-02-05 05:30:02.651 INFO 1 --- [nTaskExecutor-1] o.g.k.a.service.AsyncOrchestrator : Starting periodic book parsing
2020-02-05 05:30:02.651 INFO 1 --- [nTaskExecutor-1] o.g.komga.domain.service.LibraryScanner : Analyze all books in status: unknown
2020-02-05 05:30:02.651 INFO 1 --- [nTaskExecutor-1] o.g.komga.domain.service.LibraryScanner : Analyzed 0 books in 00:00:00.000 (virtual: 00:00:00.000)
2020-02-05 05:45:00.001 INFO 1 --- [nTaskExecutor-1] o.g.k.a.service.AsyncOrchestrator : Starting periodic libraries scan
2020-02-05 05:45:00.002 INFO 1 --- [nTaskExecutor-1] o.g.komga.domain.service.LibraryScanner : Updating library: Main, root folder: file:/books/
2020-02-05 05:45:00.002 INFO 1 --- [nTaskExecutor-1] o.g.k.domain.service.FileSystemScanner : Scanning folder: /books
2020-02-05 05:45:00.003 INFO 1 --- [nTaskExecutor-1] o.g.k.domain.service.FileSystemScanner : Supported extensions: [cbz, zip, cbr, rar, pdf]
2020-02-05 05:45:00.003 INFO 1 --- [nTaskExecutor-1] o.g.k.domain.service.FileSystemScanner : Excluded patterns: [#recycle, @eaDir]
2020-02-05 05:45:02.629 INFO 1 --- [nTaskExecutor-1] o.g.k.domain.service.FileSystemScanner : Scanned 1 series and 7449 books in 00:00:02.626
2020-02-05 05:45:02.630 INFO 1 --- [nTaskExecutor-1] o.g.komga.domain.service.LibraryScanner : Library update finished in 00:00:02.628
2020-02-05 05:45:02.630 INFO 1 --- [nTaskExecutor-1] o.g.k.a.service.AsyncOrchestrator : Starting periodic book parsing
2020-02-05 05:45:02.631 INFO 1 --- [nTaskExecutor-1] o.g.komga.domain.service.LibraryScanner : Analyze all books in status: unknown
2020-02-05 05:45:02.636 INFO 1 --- [nTaskExecutor-1] o.g.komga.domain.service.LibraryScanner : Analyzed 0 books in 00:00:00.000 (virtual: 00:00:00.000)
2020-02-05 06:00:00.000 INFO 1 --- [nTaskExecutor-1] o.g.k.a.service.AsyncOrchestrator : Starting periodic libraries scan
2020-02-05 06:00:00.001 INFO 1 --- [nTaskExecutor-1] o.g.komga.domain.service.LibraryScanner : Updating library: Main, root folder: file:/books/
2020-02-05 06:00:00.001 INFO 1 --- [nTaskExecutor-1] o.g.k.domain.service.FileSystemScanner : Scanning folder: /books
2020-02-05 06:00:00.001 INFO 1 --- [nTaskExecutor-1] o.g.k.domain.service.FileSystemScanner : Supported extensions: [cbz, zip, cbr, rar, pdf]
2020-02-05 06:00:00.001 INFO 1 --- [nTaskExecutor-1] o.g.k.domain.service.FileSystemScanner : Excluded patterns: [#recycle, @eaDir]
2020-02-05 06:00:02.585 INFO 1 --- [nTaskExecutor-1] o.g.k.domain.service.FileSystemScanner : Scanned 1 series and 7449 books in 00:00:02.584
2020-02-05 06:00:02.587 INFO 1 --- [nTaskExecutor-1] o.g.komga.domain.service.LibraryScanner : Library update finished in 00:00:02.586
2020-02-05 06:00:02.587 INFO 1 --- [nTaskExecutor-1] o.g.k.a.service.AsyncOrchestrator : Starting periodic book parsing
2020-02-05 06:00:02.587 INFO 1 --- [nTaskExecutor-1] o.g.komga.domain.service.LibraryScanner : Analyze all books in status: unknown
2020-02-05 06:00:02.587 INFO 1 --- [nTaskExecutor-1] o.g.komga.domain.service.LibraryScanner : Analyzed 0 books in 00:00:00.000 (virtual: 00:00:00.000)
2020-02-05 06:15:00.000 INFO 1 --- [nTaskExecutor-1] o.g.k.a.service.AsyncOrchestrator : Starting periodic libraries scan
2020-02-05 06:15:00.001 INFO 1 --- [nTaskExecutor-1] o.g.komga.domain.service.LibraryScanner : Updating library: Main, root folder: file:/books/
2020-02-05 06:15:00.002 INFO 1 --- [nTaskExecutor-1] o.g.k.domain.service.FileSystemScanner : Scanning folder: /books
2020-02-05 06:15:00.002 INFO 1 --- [nTaskExecutor-1] o.g.k.domain.service.FileSystemScanner : Supported extensions: [cbz, zip, cbr, rar, pdf]
2020-02-05 06:15:00.002 INFO 1 --- [nTaskExecutor-1] o.g.k.domain.service.FileSystemScanner : Excluded patterns: [#recycle, @eaDir]
2020-02-05 06:15:02.626 INFO 1 --- [nTaskExecutor-1] o.g.k.domain.service.FileSystemScanner : Scanned 1 series and 7449 books in 00:00:02.624
2020-02-05 06:15:02.628 INFO 1 --- [nTaskExecutor-1] o.g.komga.domain.service.LibraryScanner : Library update finished in 00:00:02.626
2020-02-05 06:15:02.628 INFO 1 --- [nTaskExecutor-1] o.g.k.a.service.AsyncOrchestrator : Starting periodic book parsing
2020-02-05 06:15:02.628 INFO 1 --- [nTaskExecutor-1] o.g.komga.domain.service.LibraryScanner : Analyze all books in status: unknown
2020-02-05 06:15:02.635 INFO 1 --- [nTaskExecutor-1] o.g.komga.domain.service.LibraryScanner : Analyzed 0 books in 00:00:00.000 (virtual: 00:00:00.000)
2020-02-05 06:30:00.000 INFO 1 --- [nTaskExecutor-1] o.g.k.a.service.AsyncOrchestrator : Starting periodic libraries scan
2020-02-05 06:30:00.003 INFO 1 --- [nTaskExecutor-1] o.g.komga.domain.service.LibraryScanner : Updating library: Main, root folder: file:/books/
2020-02-05 06:30:00.003 INFO 1 --- [nTaskExecutor-1] o.g.k.domain.service.FileSystemScanner : Scanning folder: /books
2020-02-05 06:30:00.003 INFO 1 --- [nTaskExecutor-1] o.g.k.domain.service.FileSystemScanner : Supported extensions: [cbz, zip, cbr, rar, pdf]
2020-02-05 06:30:00.003 INFO 1 --- [nTaskExecutor-1] o.g.k.domain.service.FileSystemScanner : Excluded patterns: [#recycle, @eaDir]
2020-02-05 06:30:02.688 INFO 1 --- [nTaskExecutor-1] o.g.k.domain.service.FileSystemScanner : Scanned 1 series and 7449 books in 00:00:02.685
2020-02-05 06:30:02.690 INFO 1 --- [nTaskExecutor-1] o.g.komga.domain.service.LibraryScanner : Library update finished in 00:00:02.687
2020-02-05 06:30:02.690 INFO 1 --- [nTaskExecutor-1] o.g.k.a.service.AsyncOrchestrator : Starting periodic book parsing
2020-02-05 06:30:02.691 INFO 1 --- [nTaskExecutor-1] o.g.komga.domain.service.LibraryScanner : Analyze all books in status: unknown
2020-02-05 06:30:02.705 INFO 1 --- [nTaskExecutor-1] o.g.komga.domain.service.LibraryScanner : Analyzed 0 books in 00:00:00.000
```

**Comment by gotson** [comment]
Thumbnails are generated as part of the book analysis, and stored in the database. They are not generated on the fly.

The web interface lazy loads books that are shown in the viewport only. Thumbnails are loaded only for books shown, and are cacheable. Even if you have a very large screen, the browser should not be loading more than 2 or 3 pages of 20 books each along with their thumbnails.

Can you try to open the Chrome dev tools and see what the network tab looks like on browsing the library?

**Comment by waz1500** [comment]
It seems to be loading every singe book in my browser(the browser scroll list is very long) and if I scroll down it will load more books.
I left it for 10 minutes and it was using about 4GB of RAM on my windows machine.  
Thumbnails are generating but I'm still unable to click on anything on the page nor can I scroll down in a quick manner. I can't even click on a book once the thumbnail is generated :/ 

I've primarily been using a test collection for my previous issues, so I've never noticed this problem in particular. 
Even the first 6 books in my collection (the top row), will re-generate its book cover upon a browser reload when If i read your response right, it should be databasing the book cover image at least. 

![komga](https://user-images.githubusercontent.com/17833884/73874282-22b11880-484b-11ea-9109-a0ca306764d9.jpg)

The logs don't seem to be giving enough debug info. Is there anything else you can recommend?

**Comment by whalehub** [comment]
I can't seem to reproduce this issue on my machine with several thousand books. The thumbnails load nearly instantly and browsing the collection is very responsive.

I'm using Chrome 79.0.3945.131 on Windows 10 and I'm running the Docker version of Komga v0.14.0 on a Debian 10 server.

**Comment by waz1500** [comment]
Thanks for giving it a go on your end whalehub. What specs are your hardware if you don't mind me asking? It looks like since I've posted the issue, a new update has come out so I'll try that and report back. I was using the stable branch and it was at 0.13.1.

**Comment by waz1500** [comment]
1. I got 304 on the first load with a white screen. It wasn't untill I clicked randomly on the white screen in which komga interface suddenly appeared but default thumbnails were there. 
![komga1e](https://user-images.githubusercontent.com/17833884/73932986-65b6ce80-48d3-11ea-978e-39923ded8c58.jpg)

2. After a while I reloaded the same library by refreshing. Same thing but thumbnails will show up. Still unable to click on anything in a timely manner. Can hover over books and see its filename etc.  
![komga2e](https://user-images.githubusercontent.com/17833884/73933159-bf1efd80-48d3-11ea-83a9-1ff55bf03ef8.jpg)

3. Selected XHR and my response times are so much slower. 
![komga3e](https://user-images.githubusercontent.com/17833884/73933268-fbeaf480-48d3-11ea-977b-05443994a6fa.jpg)

____________
![komga4e](https://user-images.githubusercontent.com/17833884/73933357-305eb080-48d4-11ea-9f65-59f382917425.jpg)

My content is a root folder called MANGA. Containing ~7500 files all which are individually zipped. Totalling about 290GB. 
 
The interface is still so slow. I will try do recreate the database with a smaller collection maybe with a 1000 .zips at a time and see at what point Komga struggles.

**Comment by gotson** [comment]
I understand better now. You should have subfolders per series, so books are bundled together. In your case you have a single series called manga with 7500 books in it!

I have never tested a series with that many books, normally the paging mechanism should be able to handle it, but there might be corner cases. The max I tried was a series with 200 books. 

Most likely your 7500 books are not part of the same series, so I suggest you should have some subfolders per manga. 

I will try to replicate the condition of your problem, I should be able to hard link all my books in a single folder to test, I might find a bottleneck somewhere.

**Comment by waz1500** [comment]
Hahaha. So for this particular issue I've been using Komga to see how it would manage my dojinishi manga content. 
There was no need to separate and create sub-folders since they were all individually zipped with it's own subfolder.  
ZIP  > [ARTIST] TITLE of folder > JPEG FILES. 

Unless I maybe combine artists etc but then it feels counter intuitive and I still would have a few thousand. 
Maybe this doesn't work since I can't really combine "series" and stuff. It's mostly just artists with a story and nothing is linked unless they're in chapters.

I use the same folder for stuff like testing lanraragi and Happypandax where their workflow is also to use one main folder with a bunch of zips in it to manage the collection. Metadata is stored within the zip. Currently Komga doesn't support it but I hope to raise the issue to support the type that I have one day. 

I will try your suggestion to ease the load by separating zips into possible artists.  
Komga works as intended for my actual comic collection which is not that extensive.

I get it if you don't want to look into this further since I doubt anyone else was trying to also use this to view their doujin content and just lumped in a bunch of files into one folder. My actual comic collection is more like the example you provided in the main page.

Thanks for helping though gotson! Much appreciated.

**Comment by gotson** [comment]
I am trying to index 4800 files in one folder to see what i get :) Will let you know.

**Comment by ticklebits** [comment]
I do exactly this with Komga as @waz1500 does. It gets really hairy as well when scripts are dumping thousands of new downloads every month into different libraries. Especially if I just dump doujins into a single folder with specific tags, much less artists or series.

Maybe libraries should have different types or settings based on what people are using it for? 

Every browser I tested chokes on thousands of files. It does load like @gotson says it does, but it takes a looong time for it to do so. So its most likely the underlying storage hardware just cant cope with tons of small files. I wonder if using SSD based storage would make it faster?

> Hahaha. So for this particular issue I've been using Komga to see how it would manage my dojinishi manga content.
> There was no need to separate and create sub-folders since they were all individually zipped with it's own subfolder.
> ZIP > [ARTIST] TITLE of folder > JPEG FILES.
> 
> Unless I maybe combine artists etc but then it feels counter intuitive and I still would have a few thousand.
> Maybe this doesn't work since I can't really combine "series" and stuff. It's mostly just artists with a story and nothing is linked unless they're in chapters. Maybe I could seperate them into folders with a tag name.
> 
> I use the same folder for stuff like testing lanraragi and Happypandax where their workflow is also to use one main folder with a bunch of zips in it to manage the collection. Metadata is stored within the zip. Currently Komga doesn't support it but I hope to raise the issue to support the type that I have one day.
> 
> I will try your suggestion to ease the load by separating zips into possible artists.
> Komga works as intended for my actual comic collection which is not that extensive.
> 
> I get it if you don't want to look into this further since I doubt anyone else was trying to also use this to view their doujin content and just lumped in a bunch of files into one folder. My actual comic collection is more like the example you provided in the main page with a max of maybe 20 .cbr files in one sub folder.
> 
> Thanks for helping though gotson! Much appreciated.

**Comment by waz1500** [comment]
I honestly think my suggestion of limiting the amount of books that appear on the page would help and probably solve this dilemma. HPX and lanraragi do this both. They also have options to display 50, 100, 150, 200 at a time etc. So this would cater towards people who have little size collections and want to display everything.

I know gotson has been working on the metadata side of things with his first few commits being simple metadata info. I'm curious though, if Komga could one day detect metadata in zips/rar's perhaps like comicinfo .xml format where fields such as "Series" and "Publisher" would be detected would there be any need to put books into separate folders?   

Komga is marked as a "free and open source comics/mangas server". But gotson might not want Komga to be used with doujinisi content which is fair enough. Then again, it supports regular official published and licensed manga like One Piece etc. Doujins are just independently published manga. 
If this is the case, separating folders for doujins in the long run just wont work. 

I'm curious as to what gotson thinks of this. Perhaps it's not really an issue. More love for the program perhaps.

**Comment by gotson** [comment]
I ran a few tests, i think the problem is really inside the browser, not on the backend. Frontend is not my strong point, but i have 2 potential trails to explore:

1. The number of DOM elements is just too big. But i did a quick search on the web and 10k elements to show shouldn't be a problem.
2. The grid view is using the intersection observer API to detect which cards are visible, and load the corresponding page accordingly. I don't know if using lots of those intersection handlers is impacting performance or not.

I need to be able to reproduce the performance drop with a simpler test case, in order to troubleshoot better.

On a side note, there is also some performance degradation in retrieving the series itself, around 10-15s to get the JSON payload instead of a few ms. I haven't had a look yet, but that shouldn't be a problem to track and fix. But that's not the main issue here.

**Comment by gotson** [comment]
Playing around with CodePen, it seems browsers don't like too many elements on the same page.

I think I need to implement pagination, what i am wondering is how to decide whether to use the current display with the full list lazily loaded, or the paginated content. 

💡 Ideas:
- automatically switch to pagination if the total number of elements is over a thresholds. How to determine the threshold? I have good results with 500 items, maybe 1000 is a good limit?
- offer a button in the toolbar to switch between full and paginated. If full is chosen and the total number of elements is over the threshold, switch back to paginated.
- remember choice of pagination per device? Or per library?

Your feedback is welcome !

**Comment by ticklebits** [comment]
Personally I would just have a single option for the whole of komga in the settings, but I could see where someone would want it per library. 

Thank you for your work @gotson!

**Comment by gotson** [comment]
> Personally I would just have a single option for the whole of komga in the settings, but I could see where someone would want it per library.

Would that be an option for the admin that would be for all the users on this Komga instance, or per user?

**Comment by waz1500** [comment]
Thanks gotson for your detailed response!
I like the second idea about having a button on the toolbar where the user could manually switch between full and paginated modes. Having Komga automatically enable the option if the number of elements is over a certain threshold. 500 items at a time would be fine for me. This would stop Komga from crashing or freezing up like in my case with large libraries.

**Comment by ticklebits** [comment]
> > Personally I would just have a single option for the whole of komga in the settings, but I could see where someone would want it per library.
> 
> Would that be an option for the admin that would be for all the users on this Komga instance, or per user?

I would say per user, as each person would have a different preference.

**Comment by gotson** [comment]
I am closing this, pagination was released in [v0.33.0](https://github.com/gotson/komga/releases/tag/v0.33.0) and should address the issue.
